pandas/_libs/interval.pyi

```diff
@@ -0,0 +1,120 @@
+from datetime import timedelta
+import numbers
+from typing import (
+    Any,
+    Generic,
+    TypeVar,
+    overload,
+)
+
+import numpy as np
+import numpy.typing as npt
+
+from pandas._libs import (
+    Timedelta,
+    Timestamp,
+)
+from pandas._typing import IntervalBound
+
+_OrderableMixinT = TypeVar(
+    "_OrderableMixinT", int, float, Timestamp, Timedelta, npt.NDArray[np.generic]
+)
+_OrderableT = TypeVar("_OrderableT", int, float, Timestamp, Timedelta)
+
+# note: mypy doesn't support overloading properties
+# based on github.com/microsoft/python-type-stubs/pull/167
+class _LengthProperty:
+    @overload
+    def __get__(self, instance: IntervalMixin[Timestamp], owner: Any) -> Timedelta: ...
+    @overload
+    def __get__(
+        self, instance: IntervalMixin[_OrderableMixinT], owner: Any
+    ) -> _OrderableMixinT: ...
+
+class IntervalMixin(Generic[_OrderableMixinT]):
+    @property
+    def closed_left(self) -> bool: ...
+    @property
+    def closed_right(self) -> bool: ...
+    @property
+    def open_left(self) -> bool: ...
+    @property
+    def open_right(self) -> bool: ...
+    @property
+    def mid(self) -> _OrderableT: ...
+    length: _LengthProperty
+    @property
+    def is_empty(self) -> bool: ...
+    def _check_closed_matches(self, other: IntervalMixin, name: str = ...) -> None: ...
+
+class Interval(IntervalMixin[_OrderableT]):
+    def __init__(
+        self,
+        left: _OrderableT,
+        right: _OrderableT,
+        closed: IntervalBound = ...,
+    ) -> None: ...
+    @property
+    def closed(self) -> str: ...
+    @property
+    def left(self) -> _OrderableT: ...
+    @property
+    def right(self) -> _OrderableT: ...
+    def __str__(self) -> str: ...
+    # TODO: could return Interval with different type
+    def __add__(
+        self, y: numbers.Number | np.timedelta64 | timedelta
+    ) -> Interval[_OrderableT]: ...
+    def __radd__(
+        self, y: numbers.Number | np.timedelta64 | timedelta
+    ) -> Interval[_OrderableT]: ...
+    def __sub__(
+        self, y: numbers.Number | np.timedelta64 | timedelta
+    ) -> Interval[_OrderableT]: ...
+    def __mul__(self, y: numbers.Number) -> Interval[_OrderableT]: ...
+    def __rmul__(self, y: numbers.Number) -> Interval[_OrderableT]: ...
+    def __truediv__(self, y: numbers.Number) -> Interval[_OrderableT]: ...
+    def __floordiv__(self, y: numbers.Number) -> Interval[_OrderableT]: ...
+    def __hash__(self) -> int: ...
+    def __contains__(self: Interval[_OrderableT], key: _OrderableT) -> bool: ...
```

Comment:
I think you have to be explicit here about the 4 different types.  You can't have `__contains__()` support testing an integer inside a `Timestamp` interval.  See PR https://github.com/microsoft/python-type-stubs/pull/167

Reply:
The two `_OrderableT` bind to the same type simultaneously. 

```py
i2 = pd.Interval(10, 20)
i2.__contains__(4) # ok
i2.__contains__(4.0) # error: Unsupported operand types for in ("float" and "Interval[int]")  [operator]

i3 = pd.Interval(13.2, 19.5)
i3.__contains__(4) # ok
i3.__contains__(4.0) # ok
i3.__contains__(pd.Timestamp(0)) # error: Unsupported operand types for in ("Timestamp" and "Interval[float]")  [operator]
```

Thank you for your feedback! I will integrate it over the next days. If you prefer, I can also leave the operators as-is and leave it to you - doesn't feel great copy&pasting your code ;)

Reply:
I think that `i2.__contains__(4.0)` should be allowed.  If you have an interval based on integers, you can test whether a float is inside.